### PR TITLE
fix(platform-ios): fix `sourceDir` detection

### LINKED ISSUE #1435: run-ios bug: Try/catch removed from `warnAboutManuallyLinkedLibs.ts` causing exceptions to propagate/crash
## Environment

```
System:
  OS: macOS 11.4
  CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  Memory: 921.95 MB / 16.00 GB
  Shell: 3.2.57 - /bin/bash
Binaries:
  Node: 16.4.0 - /usr/local/bin/node
  Yarn: 1.22.10 - /usr/local/bin/yarn
  npm: 7.18.1 - /usr/local/bin/npm
  Watchman: 4.9.0 - /usr/local/bin/watchman
Managers:
  CocoaPods: 1.10.1 - /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
  Android SDK:
    API Levels: 28, 29, 30
    Build Tools: 28.0.3, 29.0.3, 30.0.1
    System Images: android-30 | Google APIs Intel x86 Atom
    Android NDK: Not Found
IDEs:
  Android Studio: 4.1.2 4.1.2
  Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
Languages:
  Java: 11.0.1 - /usr/bin/javac
npmPackages:
  @react-native-community/cli: Not Found
  react: 17.0.1 => 17.0.1 
  react-native: 0.64.1 => 0.64.1 
  react-native-macos: ^0.63.0 => 0.63.36 
npmGlobalPackages:
  *react-native*: Not Found
```

## Description

This refers to https://github.com/microsoft/react-native-test-app/issues/375

I was working on moving a community module to use [React Native Test App](https://github.com/microsoft/react-native-test-app/) to manage the test app of a community module (https://github.com/react-native-menu/menu), and ran into an error running `run-ios`. After investigation, it seems the issue happened upstream beginning in https://github.com/react-native-community/cli/pull/1270 .

A try-catch was removed from `packages/platform-ios/src/link/warnAboutManuallyLinkedLibs.ts`, causing the exception to bubble up and crash. This seems like an error, because an Android fix shouldn't have broken something downstream for iOS.

## Reproducible Demo

A little bit contrived since I reproed this using a downstream module, but:
```
git clone git@github.com:react-native-menu/menu.git
cd menu
mv example/ example2/ # This is to just move the existing test app out of the way temporarily
yarn add react-native-test-app --dev
yarn init-test-app
# The options I chose
# ✔ What is the name of your test app? … MenuExample
# ✔ Which platforms do you need test apps for? › Android, iOS
# ✔ Where should we create the new project? … example
yarn
yarn build:ios
pod install --project-directory=ios
yarn ios
```

### PR DESCRIPTION
**Pull request by tido64**
Summary:
---------

This is a minimal changes version of #1436. After speaking with @grabbou, I learned that most of the props in `config` are going away, so we agreed to make as little changes as possible and instead focus on `next`.

Resolves #1435
Resolves https://github.com/microsoft/react-native-test-app/issues/375

Test Plan:
----------

Curren tests should pass:
```
 PASS   unit  packages/platform-android/src/link/patches/__tests__/makeSettingsPatch.test.ts
 PASS   unit  packages/platform-ios/src/commands/runIOS/__tests__/findMatchingSimulator.test.ts
 PASS   unit  packages/platform-android/src/config/__tests__/getProjectConfig.test.ts
 PASS   unit  packages/cli/src/commands/init/__tests__/template.test.ts
 PASS   unit  packages/cli/src/tools/__tests__/packageManager-test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getHeaderSearchPath.test.ts
 PASS   unit  packages/cli/src/commands/link/__tests__/link.test.ts
 PASS   unit  packages/platform-android/src/commands/runAndroid/__tests__/runOnAllDevices.test.ts
 PASS   unit  packages/cli/src/tools/windows/__tests__/androidWinHelpers.test.ts
 PASS   unit  packages/cli/src/commands/upgrade/__tests__/upgrade.test.ts (5.274 s)
 PASS   unit  packages/cli/src/commands/bundle/__tests__/getAssetDestPathAndroid-test.ts
 PASS   unit  packages/platform-ios/src/config/__tests__/findProject.test.ts
 PASS   unit  packages/cli/src/commands/init/__tests__/editTemplate.test.ts (6.21 s)
 PASS   unit  packages/platform-android/src/config/__tests__/findPackageClassName.test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/makeBuildPatch.test.ts
 PASS   unit  packages/platform-ios/src/commands/runIOS/__tests__/parseXctraceIOSDevicesList.test.ts
 PASS   unit  packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/removePodEntry.test.ts
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/isInstalled.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/writePlist.test.ts
 PASS   unit  packages/cli/src/tools/__tests__/copyFiles.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/createGroup.test.ts
 PASS   unit  packages/platform-android/src/config/__tests__/getDependencyConfig.test.ts
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/findLineToAddPod.test.ts
 PASS   unit  packages/platform-ios/src/config/__tests__/findPodspec.test.ts
 PASS   unit  packages/cli/src/commands/doctor/healthchecks/__tests__/androidHomeEnvVariable.test.ts
 PASS   unit  packages/cli/src/commands/bundle/__tests__/getAssetDestPathIOS-test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/addSharedLibraries.test.ts
 PASS   unit  packages/platform-ios/src/commands/runIOS/__tests__/findXcodeProject.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getGroup.test.ts
 PASS   unit  packages/platform-ios/src/config/__tests__/getProjectConfig.test.ts
 PASS   unit  packages/cli/src/commands/bundle/__tests__/filterPlatformAssetScales-test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/isInstalled.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/removeProjectFromProject.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/removeSharedLibrary.test.ts
 PASS   unit  packages/cli/src/tools/config/__tests__/findAssets-test.ts
 PASS   unit  packages/cli/src/tools/config/__tests__/index-test.ts (12.208 s)
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/findMarkedLinesInPodfile.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getHeadersInFolder.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/removeProjectFromLibraries.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/mapHeaderSearchPaths.test.ts
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/findPodTargetLine.test.ts
 PASS   unit  packages/platform-android/src/link/__tests__/isInstalled.test.ts
 PASS   unit  packages/platform-android/src/config/__tests__/readManifest.test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/makePackagePatch.test.ts
 PASS   unit  packages/cli/src/commands/init/__tests__/validate.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/hasLibraryImported.test.ts
 PASS   unit  packages/cli/src/tools/config/__tests__/findProjectRoot-test.ts
 PASS   unit  packages/cli/src/commands/doctor/healthchecks/__tests__/androidStudio.test.ts (10.519 s)
 PASS   unit  packages/platform-ios/src/link/__tests__/addProjectToLibraries.test.ts
 PASS   unit  packages/cli/src/commands/doctor/healthchecks/__tests__/jdk.test.ts (13.899 s)
 PASS   unit  packages/cli/src/commands/doctor/healthchecks/__tests__/androidSDK.test.ts (13.944 s)
 PASS   unit  packages/platform-ios/src/link/__tests__/addFileToProject.test.ts
 PASS   unit  packages/platform-android/src/config/__tests__/findAndroidDir.test.ts
 PASS   unit  packages/platform-android/src/config/__tests__/findManifest.test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/makeImportPatch.test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/makeStringsPatch.test.ts
 PASS   unit  packages/platform-ios/src/link-pods/__tests__/getDependenciesFromPodfileLock.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getTargets.test.ts
 PASS   unit  packages/cli/src/commands/doctor/healthchecks/__tests__/androidNDK.test.ts (11.249 s)
 PASS   unit  packages/platform-ios/src/link/__tests__/getBuildProperty.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getPlistPath.test.ts
 PASS   unit  packages/cli/src/tools/config/__tests__/findDependencies-test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/applyParams.test.ts
 PASS   unit  packages/tools/src/__tests__/groupFilesByType.test.ts
 PASS   unit  packages/platform-android/src/link/patches/__tests__/normalizeProjectName.test.ts
 PASS   unit  jest/__tests__/replaceProjectRootInOutput.test.ts
 PASS   unit  packages/platform-ios/src/config/__tests__/findPodfilePath.test.ts
 PASS   unit  packages/platform-ios/src/link/__tests__/getPlist.test.ts

 PASS   unit  packages/cli/src/tools/__tests__/loadMetroConfig-test.ts
 PASS   unit  packages/cli/src/commands/link/__tests__/makeHook.test.ts
1;
 PASS   e2e  __e2e__/unknown.test.ts
 PASS   e2e  __e2e__/default.test.ts
 PASS   e2e  __e2e__/uninstall.test.ts (11.808 s)
 PASS   unit  packages/cli/src/commands/info/__tests__/info.test.ts (8.752 s)
 PASS   e2e  __e2e__/install.test.ts (10.138 s)
 PASS   e2e  __e2e__/init.test.ts (30.857 s)
 PASS   e2e  __e2e__/config.test.ts (135.075 s)
 PASS   e2e  __e2e__/root.test.ts (138.565 s)

Test Suites: 2 skipped, 79 passed, 79 of 81 total
Tests:       10 skipped, 2 todo, 287 passed, 299 total
Snapshots:   28 passed, 28 total
Time:        147.844 s
Ran all test suites in 2 projects.
```

### CODE REVIEW

**Comment by thymikee** [comment]
@grabbou wdyt?

**Comment by kelset** [comment]
@thymikee is this blocked on @grabbou, or could we get it in? It's been open for a while 😓

**Comment by thymikee** [comment]
I didn't have the time to properly review this yet, so would still like @grabbou or @Esemesek to take a look

File: packages/platform-ios/src/config/index.ts [hunk 1]
```diff
@@ -48,13 +48,18 @@ export function projectConfig(folder: string, userConfig: IOSProjectParams) {
   }
 
   const projectPath = path.join(folder, project);
-  const sourceDir = path.dirname(projectPath);
+  const podfile = findPodfilePath(folder, projectPath);
+
+  // This is a temporary fix for #1435. In certain repos, the Xcode project can
+  // be generated by a tool. The only file that we can assume to exist on disk
+  // is `Podfile`.
+  const sourceDir = podfile ? path.dirname(podfile) : path.dirname(projectPath);
```
Comment (grabbou) [inline packages/platform-ios/src/config/index.ts:56]: I would not change this one to be honest, `sourceDir` is not needed for `autolinking`, so there's no need to adjust it for cases when `xcodeproj` doesn't exist. It is needed for legacy `link` command and in the case of a missing `xcodeproj`, that would fail anyway.

**Comment by grabbou** [comment]
I am going to close this PR and let's continue this discussion in the https://github.com/react-native-community/cli/issues/1435 for the most optimal solution.

TL;DR - I could not reproduce your issues on `master`, it may be that we have done some other changes in the meantime (or you did them) that made this work out of the box.

In case I misunderstood anything, I will be happy to reopen or discuss this once again on Discord. 

Thank you!

**Comment by tido64** [comment]
> That said, I am hesitant to add additional complexity to something that already works, as well as change the way we resolve a Podfile - at the risk of breaking this for our other users.

But it's not fundamentally changing how you're resolving a Podfile today. The logic is still the same for the success case. Only in the case of it failing, are we resorting to a glob search. I don't understand how that is a risk? If it was failing before, it wouldn't be any worse with the current change.

File: packages/platform-ios/src/config/index.ts (hunk 1 above)
Comment (tido64) [reply packages/platform-ios/src/config/index.ts:56]: It looks like it _is_ being used for autolinking:
https://github.com/react-native-community/cli/blob/44f541332ee57e8a30feb3ca24479ab292149dd9/packages/platform-ios/native_modules.rb#L40

Without this change, we hit errors similar to this when running `pod install`:
```
[!] No podspec found for `ReactTestApp-DevSupport` in `..`
```

**Comment by grabbou** [comment]
Reopening as it looks like that’s our only way to go in this case for time being.

File: packages/platform-ios/src/config/index.ts (hunk 1 above)
Comment (grabbou) [reply packages/platform-ios/src/config/index.ts:56]: You’re right 😅

**Review by grabbou — APPROVED** [review]

**Comment by grabbou** [comment]
After taking another pass and your explanations @tido64, it looks good 👍

**Comment by thymikee** [comment]
@tido64 published `@react-native-community/cli-platform-ios@5.0.2` with this fix as well. Thanks!

**Comment by tido64** [comment]
Thanks for backporting it, @thymikee. Did it get published? I don't see it in the feed yet: https://www.npmjs.com/package/@react-native-community/cli?activeTab=versions

**Comment by joshchoo** [comment]
Hi @tido64,

I'm facing a problem with podspec path resolution on one of the packages I'm using: react-native-emarsys-wrapper.
The package doesn't use auto-linking, and it defines a sample Podfile in [sample/ios/Podfile](https://github.com/emartech/react-native-emarsys-sdk/blob/371311e/sample/ios/Podfile).

This change causes sourceDir to resolve to `sample/ios` instead of `ios` because of the location of the sample Podfile. Consequently, `native_modules.rb` can't find the podspec file and logs the warning "_use_native_modules! skipped the react-native dependency react-native-emarsys-wrapper. No podspec file was found_".

I'm currently working around this by patching react-native-emarsys-wrapper to remove sample/ios/Podfile to avoid incorrect sourceDir resolution. Is there a better way to resolve sourceDir, such as editing `react-native.config.js`, to fix it? Or should I wait for this temporary fix to be removed?

**Environment**
- React Native: 0.66.0

**Comment by tido64** [comment]
@joshchoo: You can explicitly disable it in your `react-native.config.js`: https://github.com/react-native-community/cli/blob/master/docs/autolinking.md#how-can-i-disable-autolinking-for-unsupported-library

This way you don't have to patch the package.

Is the `sample` folder an essential part of the package? Otherwise, I would exclude it from the published package. As far as I know, the "next" version that will make this temporary fix unnecessary is being worked on, but I don't have any other information.
